File: articles/calculus-for-systems-modeling/articles/related-rates-and-interdependent-motion/fortran/related_rates_audit.f90
program related_rates_audit
  implicit none
  real(8), dimension(5) :: ts
  real(8) :: t,h,hr,v,structural,inferred
  integer :: i
  ts = (/0.0d0,5.0d0,10.0d0,20.0d0,40.0d0/)
  print '(A)', 'time height height_rate volume structural_derivative inferred_volume_rate'
  do i=1,size(ts)
    t=ts(i); h=height_path(t); hr=height_rate(t); v=volume(h); structural=d_volume_d_height(h); inferred=structural*hr
    print '(F8.3,1X,F14.6,1X,F14.6,1X,F14.6,1X,F14.6,1X,F14.6)', t,h,hr,v,structural,inferred
  end do
contains
  real(8) function volume(h)
    real(8), intent(in) :: h
    volume=12.0d0*h*h
  end function
  real(8) function d_volume_d_height(h)
    real(8), intent(in) :: h
    d_volume_d_height=24.0d0*h
  end function
  real(8) function height_path(t)
    real(8), intent(in) :: t
    height_path=2.0d0 + 0.08d0*t
  end function
  real(8) function height_rate(t)
    real(8), intent(in) :: t
    height_rate=0.08d0
  end function
end program

File: articles/calculus-for-systems-modeling/articles/related-rates-and-interdependent-motion/r/related_rates_audit.R
volume <- function(height, shape_coefficient = 12) {
  shape_coefficient * height^2
}

d_volume_d_height <- function(height, shape_coefficient = 12) {
  2 * shape_coefficient * height
}

height_path <- function(time) {
  2 + 0.08 * time
}

height_rate <- function(time) {
  0.08
}

finite_difference_volume_rate <- function(time, h = 1e-4) {
  (volume(height_path(time + h)) - volume(height_path(time - h))) / (2 * h)
}

audit_time <- function(time) {
  current_height <- height_path(time)
  current_height_rate <- height_rate(time)
  current_volume <- volume(current_height)
  structural <- d_volume_d_height(current_height)
  inferred_rate <- structural * current_height_rate
  fd <- finite_difference_volume_rate(time)
  error <- abs(inferred_rate - fd)

  warning <- ""
  if (current_height <= 0) {
    warning <- "height outside physical domain"
  } else if (error > 1e-5) {
    warning <- "finite-difference check differs from related-rate calculation"
  }

  data.frame(
    time = time,
    height = current_height,
    height_rate = current_height_rate,
    volume = current_volume,
    structural_derivative = structural,
    inferred_volume_rate = inferred_rate,
    finite_difference_check = fd,
    absolute_error = error,
    warning = warning
  )
}

results <- do.call(rbind, lapply(c(0, 5, 10, 20, 40), audit_time))
dir.create("outputs/tables", recursive = TRUE, showWarnings = FALSE)
write.csv(results, "outputs/tables/r_related_rates_audit.csv", row.names = FALSE)
print(results)
